### ISSUE TITLE
Add assertion helpers assertLogEvent/assertLogEvents for testing log events

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7934]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)| @eeshangarg reported|
|-|-|
|Trac ID|trac#7934|
|Type|enhancement|
|Created|2015-06-11 08:38:51Z|

Attachments:

* [feature_7934.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/838/838ce8f3fe194fa670d144d738c65d29275f2c1c/a211a8e652e026a1439c5a4279acdb748113cb10.patch) (13944 bytes) - added by eeshangarg on 2015-06-11 08:49:10Z - 
* [feature_7934_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/838/838ce8f3fe194fa670d144d738c65d29275f2c1c/20eb17b7e548df187a7dd83f06102d2095ee58f2.patch) (12010 bytes) - added by eeshangarg on 2015-06-17 12:19:48Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7934 7934
type__enhancement enhancement
reporter__eeshangarg eeshangarg
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__None None
time__1434011931714763 1434011931714763
changetime__1477953986010676 1477953986010676
version__None None
owner__eeshangarg eeshangarg
cc__eeshangarg
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_1"></a>|@eeshangarg removed owner|
|-|-|

Hi!

I just submitted a patch that adds two assertion helpers, assertLogEvent and assertLogEvents. Here are a few things that I would like you to know:

    1. I put the assertion helpers somewhere private for now so that we can first try them out before committing to them and then make them public when appropriate.

    2. Both assertion helpers take a testCase argument which is the TestCase object that wants to make use of the helpers. Note that I initially implemented these helpers as part of a specialized TestCase subclass but I dropped this idea, and I instead reimplemented the helpers as stand-alone functions that accept a testCase parameter in order to call those assert* methods.

    3. Apart from testCase, event, and kwargs, assertLogEvent takes 5 arguments out of which two are somewhat optional and have a default value of None. The reason is that logger, level and namespace should *always* be there as these are what makes up a proper log event in the first place, whereas source and format cannot be completely absent, but can be None; in which case if we expect them to be None, we should not need to explicitly pass None as their argument, the default parameter value should do that for us.

    4. I separated the assertion logic in such a way that assertLogEvent makes only one assertion; so that we know when *everything* goes wrong, rather than just knowing about one of half a dozen assertions failing (which was the case in my previous implementation, which simply grouped together a few calls to different assert* methods). So we build up an expectedEvent dict and assertEqual that with the event dict being tested.

5. A few things worth noticing - log_failure and log_time are not tested. I had no idea how to test log_time, and if I am not mistaken, I think log_time is not even tested in the tests for twisted.logger.Logger itself. As for log_failure, one thing that I learned was that exceptions differ and it can't be asserted that two exceptions are equal unless they are both reproduced in the same context. Plus, Logger.failure is used to log unexpected failures, if they were expected, they would either be handled or logged using Logger.error. So, it seems to me that there's little point in testing log_failure, not to mention that it is difficult to reproduce and test for an exception that we don't expect to see. However, I can be wrong about this.

6. One caveat of asserting the equality of two dictionaries is that if the dicts being compared are large, then even in the smallest of failure cases, the traceback message diff is truncated. This can be easily remedied by setting self.maxDiff to None for that particular testCase for which we want to see the full diff. I have quadruple-checked and if you set self.maxDiff to None within the scope of a test_* method, then that change holds only for that particular test method and is reset to the default value when the test runner moves on to the next test method. Now when a test method that uses assertLogEvent(s) is run, if an assertion fails, the traceback diff is quite useful and informative.

7. assertLogEvents takes a numOfExpectedEvents argument, which is optional. Irrespective of whether this argument is passed or not, the helper will still assert the length of actualEvents and expectedEvents as I thought that we should know when both the implicit and the explicit assertions about the length of the event dicts being compared fail.

8. Lastly, in assertLogEvents, an inner for loop strips 'log_' from the start of every key in expectedEvent so that expectedDict can be properly unpacked as arguments to assertLogEvent. In order to get rid of this inner for loop, we would have to modify assertLogEvent's parameters so that the Logger related arguments start with 'log_' (change argument names like format to log_format, level to log_level and so on; which would not conform to our coding standard).

Phew! That was pretty much all I wanted to say. Anyways, if this implementation sucks, I would be more than happy to start from scratch. :-)

Thanks to glyph, tomprince and adiroiban for their help!

Regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban set owner to @eeshangarg|
|-|-|

Hi,

Here is just a quick review... as I don't have time/mood to do a full review.

Also I know / care little about the logging system so it is preferred if the final review will be done by someone who has implemented the new logging system.

I do care about reviews getting a quick feedback.. so here I go.

Hope it will give some good feedback and future reviews will take less.

Thanks!

------


1. We are in the test part of the Twisted code and here everything is private and we can change it without depreacation. Feel free not to worry about public/private stuff. Some things which should be really private can be proctes using `__varName` .... I know that wsanches prefer to have private stuff in the test code... feel free to use whatever you want / feel suitable :)

2. I prefer to have the test helpers implentes as a mixin ... assertion Mixin

```
class LogAssertionMixin(object):

    def assertLogEvent(self, event, etc):
        """
        Use self instead of testCase.
        """

    def assertLogEvents(self, actualEvents, expectedEvents, etc):
        pass



class SomeRealTests(unit.TestCase, LogAssertionMixin):

    def test_something(self):
        # bla
        self.assertLogEvent(event)

```

3. OK. I assume that by starting to use these assertions in other tests we will get a sense of how the API should look like

4. Again, usage will tell if the default dict assertion error is good enought to help in this case.

5. Time is usually tested by injecting a rigged timer/clock. I don't know what to say about log_failure. I think that current implementation is ok and once somebody has a need to check some log_failure part we can consider how to test it.

6. OK. Many thanks for the note.

7. Don't know what to say... I feel that we don't need it. 

8. I would say that is ok if assertLogEvent arguments start with `log_`

-----

`newKeyName = key.split('_')[1]` might work for `log_something` but it will not work for `log_something_else`... you just just do `newKeyName = key[4:]`... but I hope we can get rid of it.

I like to keep testing code simple, at the expense of verbosity as we don't have tests for the tests ... of the test and simple code should be easier to inspect and maintain.

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_3"></a>|@eeshangarg removed owner|
|-|-|

Hi!

Thanks for the review, adiroiban! I just added another patch (feature_7934_2.patch​) for this. I modified assertLogEvent to accept parameters that start with 'log_' (twistedchecker will complain about this!), so that we can get rid of that inner for loop in assertLogEvents in my previous patch. Plus, now, assertLogEvents accepts no numOfExpectedEvents parameter. These two changes also necessitated removing two redundant tests and modifying existing tests to pass correctly named keyword arguments to assertLogEvent (for example, log_source=sourceObject, instead of just source=sourceObject).

I did not implement the helpers as part of a mixin yet. Before I go ahead and submit a patch that does that, I have some concerns that I would like to share:

   1. If I want to modify twisted.internet.protocol.Factory to use the new logging system, that would also involve adding tests that test the new logging statements (see [ticket:7897]); then the test class for twisted.internet.protocol.Factory would have to inherit from our LogAssertionMixin.

   2. If we are to port the Twisted codebase to the new logging system and start using it exclusively, if I am not mistaken, we would have to repeat the process outlined above in 1 for every component of Twisted that has some logging statements. I am guessing that if we do that, we'll get to a point where every test class that tests something that logs something will have to inherit from LogAssertionMixin (which is A LOT of test classes!) and we will end up with a significantly large majority of test classes in Twisted inheriting from this mixin. I am doubtful as to whether this is a good thing or not and in my humble opinion, this definitely warrants a somewhat longer discussion. :-)

Regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_4"></a>|@wsanchez commented|
|-|-|

It would be useful to have an example of how to use this somewhere… maybe in the testing howto?

http://twistedmatrix.com/documents/current/core/howto/testing.html

Can't all of the `log_` arguments to `assertLogEvent` be covered by `kwargs`?

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_5"></a>|@eeshangarg commented|
|-|-|

Thanks for your comments, wsanchez!

I would be happy to file and work on a follow-up ticket to have an example of how to use the assertion helpers in the testing how-to once this ticket is merged and closed.

About using kwargs to represent all the log_ arguments to assertLogEvent, yes that could be done; I implemented it this way because:

a) It is similar to the API that twisted.logger.Logger's existing log events emitting methods use, where there are a couple of arguments that should always be there (log_logger, log_level, log_namespace, these arguments are what makes up a log event dict in the first place and should always be there and should always be tested), along with things that are always present but may be None, and hence could be passed as optional keyword arguments with a default value of None (log_source and log_format. Note that when we want to test that these are None, we don't have to pass an argument at all, the default value will take care of that for us. Not to mention that if we use kwargs for everything, we won't be able to specify a default value of None for these arguments in the function signature.). At last we have kwargs, which is just a way for the code that logs something to pass arbitrary (and arbitrarily named) keyword arguments that are related to the event being logged.

b) While I agree that using kwargs would result in a much more concise implementation; since keyword arguments passed via kwargs can be arbitrarily named, I think the current function signature is much more explicit about what it expects rather than depending on a kwargs dict being structured in a particular way with things that must be there, things that may or may not be None, followed by an arbitrary amount of arbitrarily named keyword arguments that the code that is logging the event thinks are necessary to describe the event.

If you disagree with any of the above and feel that there is a better way, I would be happy to work on this further and implement the changes that you put forth. :-)

Thanks & regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to @eeshangarg|
|-|-|

Thanks for following up with this code, eeshangarg.  I have a bunch of feedback.

    1. I think you should take wsanchez's suggestion as something to do beforehand.  Sometimes, when writing documentation for a feature like this, you will discover that it makes a lot more sense to factor it some other way, and it will change the implementation quite a bit.

    2. Since this is a follow-up to [#7897](https://github.com/twisted/twisted/issues/7897), I think this branch should also at least include a re-factoring of `test_doStartLoggingStatement` and `test_doStopLoggingStatement` to use it.  If those two tests aren't made more readable then there isn't much of a point to all this code :).

    3. Given that this functionality is in a double-private module (private because it's in a `test` package, private again because it starts with an underscore) I think it would be inappropriate to document it with a `.feature` NEWS file.  There are two possible solutions to this: make it private functionality only for use within Twisted to try it out, and make it a `.misc` NEWS file, later publishing it, or polish it enough to make it public.  Right now, how to test log messages is really unclear even with the fairly nice documentation for `twisted.logger`.  However, I'll leave this up to you as the author.

    4. `_getNamespace` is a poor re-implementation of `twisted.python.reflect.fullyQualifiedName`.  Please delete it and replace it with that.

    5. It occurs to me that the `self.logger.observer = self.events.append` pattern is super clear if you understand exactly how the new logging system works, but you have to fully internalize some descriptor magic to get it and it isn't documented anywhere.  I think we may want to add some contextmanagers for this: `with capturingEventsFrom(someObject.log) as captured:`, for example, would provide a nice public entry point.  I think you might have discovered this was somewhat awkward to describe while writing documentation, so yet another reason to do that :-).

    6. For what it's worth, inheritance is bad (adiroiban clearly needs to go watch [https://www.youtube.com/watch?v=3MNVP9-hglc the talk] again) and I'm glad you didn't do this as a mixin :-).

Although the code is in many ways quite polished, I think it could still do with a round of re-review so please address the points above and resubmit.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_7"></a>|@wsanchez commented|
|-|-|

I had selfish reviewer reasons for wanting documentation, which was that I didn't quite know how this was meant to be used, so I couldn't really do a good review without studying it for longer than I've had available time for lately.

I didn't notice the double-private-ness which makes the location of the docs I suggested a bad one, but an example in a doc doctoring would be handy.

glyph's suggestion #5 sounds like a great usability win.

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_8"></a>|@eeshangarg removed owner|
|-|-|

Just submitted a PR for this, <https://github.com/twisted/twisted/pull/520>.

Thanks!
Eeshan

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to @eeshangarg|
|-|-|

Hi eeshangarg,

Thanks for resubmitting this. I hope we can drive this through to completeness. :).

I'd love to address the more substantive parts of your questions here, but I've skipped this review the last few times I came across it, because the tests are failing.  Can you please update your branch so that it merges cleanly, then ensure that the tests are passing (it looks like the issue is your code isn't python 3 compatible).

If you can get it into shape where it _could_ be merged, it'll make it a lot more likely that I'll have a look at the code.  Have a look at the travis-CI and appveyor failures on the pull request for more info about what needs to be fixed.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_10"></a>|@markrwilliams commented|
|-|-|

Hi eeshangarg,

I just finished a review [in Github](https://github.com/twisted/twisted/pull/520#pullrequestreview-3996657).  Please note that until further notice (that is, until I actually finish the supporting infrastructure for it), Github isn't the source of record for patches that need review.  Instead, you should update the Trac ticket (this very one for that reviewed Github PR) so that it includes the "review" keyword.  You'll know you've got it when it shows up at http://twisted.reviews.

I'll do that now so that others will know they should take a look at what I consider a very important patch.

Thanks so much for your work!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

It looks like, workflow-wise, this is awaiting feedback from eeshan, so it _shouldn't_ have the review keyword on it.  Perhaps a more appropriate way to say "hey everybody look at this" would be to just drop a message to the mailing list, rather than putting it into an ambiguous workflow state?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

Since there were no objections…
